File: cpps/06/fun_vs_pointer/fun_pointer.py
#! /usr/bin/env python3
# -*-coding:utf-8 -*-
# @Time    : 2021/01/07 18:53:21
# @Author  : che
# @Email   : ch1huizong@gmail.com


def max(a, b):
    return a > b and a or b


def min(a, b):
    return a < b and a or b


def sum(a, b):
    return a + b


def compute(a, b, f):
    return f(a, b)


if __name__ == "__main__":
    a = int(input("请输入整数a: "))
    b = int(input("请输入整数b: "))
    print("Max a and b is ", compute(a, b, max))
    print("Min a and b is ", compute(a, b, min))
    print("Sum a and b is ", compute(a, b, sum))

File: cpps/06/fun_vs_pointer/fun_pointer.cpp
#include <iostream>
// T: 函数指针

int max(int a, int b) { return ((a > b) ? a : b); }

int min(int a, int b) { return ((a < b) ? a : b); }

int sum(int a, int b) { return a + b; }

int compute(int a, int b, int (*f)(int a, int b)) {  // 定义函数指针
  // return (*f)(a, b);
  return f(a, b);
}

int main(int argc, char* argv[]) {
  int a, b, res;
  std::cout << "请输入整数a: ";
  std::cin >> a;
  std::cout << "请输入整数b: ";
  std::cin >> b;

  res = compute(a, b, &max);  // 可以直接传函数名max
  std::cout << "Max of " << a << " and " << b << " is " << res << std::endl;

  res = compute(a, b, &min);
  std::cout << "Min of " << a << " and " << b << " is " << res << std::endl;

  res = compute(a, b, &sum);
  std::cout << "Sum of " << a << " and " << b << " is " << res << std::endl;

  return 0;
}
